App/Program.fs

```diff
@@ -22,46 +23,50 @@ let writeUrl =
 let readUrl =
   Environment.tryGetEnv "MESSAGE_DB_REPLICA_URL" |> Option.defaultValue writeUrl
 
-let connection = MessageDbConnector(writeUrl, readUrl).Establish()
-let context = MessageDbContext(connection)
+let connection = MessageDbClient(writeUrl, readUrl)
+let context = MessageDbContext(connection, 100)
 let caching = CachingStrategy.SlidingWindow(cache, TimeSpan.FromMinutes(20))
 
 let service =
   MessageDbCategory(context, Invoice.Events.codec, Invoice.Fold.fold, Invoice.Fold.initial, caching)
   |> Equinox.Decider.resolve log
   |> Invoice.create
 
-let builder = WebApplication.CreateBuilder()
-let app = builder.Build()
-
-let raiseInvoice body =
-  task {
-    let id = Guid.NewGuid() |> Invoice.InvoiceId.ofGuid
-    do! service.Raise(id, body)
-    return id
-  }
-
-app.MapPost("/", Func<_, _>(raiseInvoice)) |> ignore
-
-let finalizeInvoice id =
-  task {
-    do! service.Finalize(id)
-    return "OK"
-  }
-
-app.MapPost("/{id}/finalize", Func<_, _>(finalizeInvoice)) |> ignore
+let numberService =
+  let access = AccessStrategy.AdjacentSnapshots(InvoiceNumbering.Fold.snapshotEventType, InvoiceNumbering.Fold.toSnapshot)
+  MessageDbCategory(context, InvoiceNumbering.Events.codec, InvoiceNumbering.Fold.fold, InvoiceNumbering.Fold.initial, caching, access)
+  |> Equinox.Decider.resolve log
+  |> InvoiceNumbering.create
 
-let recordPayment id payment =
-  task {
-    do! service.RecordPayment(id, payment)
-    return "OK"
-  }
+let reactorService = InvoiceNumberingReactor.Service(numberService, service)
+let checkpointService =
+  MessageDbCategory(context, CheckpointStore.Events.codec, CheckpointStore.Fold.fold, CheckpointStore.Fold.initial, caching, access = AccessStrategy.LatestKnownEvent)
+  |> Equinox.Decider.resolve log
+  |> CheckpointStore.create
 
-app.MapPost("/{id}/record-payment", Func<_, _, _>(recordPayment)) |> ignore
 
-let readInvoice id =
-  task { return! service.ReadInvoice(id) }
+let builder = WebApplication.CreateBuilder()
+builder.Services.AddHostedService(fun _ -> new HostedReactor.HostedReactor(readUrl, log, reactorService, checkpointService)) |> ignore
```

Comment:
maybe pass relevant inputs to a hostedreactor.create, and let it create the ReactorService ?

Reply:
Ended up adding a `Services` type to encapsulate the creation of these services. I think that's a lot clearer, thanks for the suggestion